master/buildbot/reporters/generators/comment.py

```diff
@@ -0,0 +1,154 @@
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+import bisect
+import json
+
+from twisted.internet import defer
+from zope.interface import implementer
+
+from buildbot import interfaces
+from buildbot.reporters import utils
+from buildbot.reporters.message import MessageFormatterFunction
+
+from .utils import BuildStatusGeneratorMixin
+
+
+@implementer(interfaces.IReportGenerator)
+class BuildCodeIssueCommentsGenerator(BuildStatusGeneratorMixin):
+
+    wanted_event_keys = [
+        ('builds', None, 'finished'),
+    ]
+
+    compare_attrs = ['formatter']
+
+    def __init__(self, diffinfo_data_name=None, tags=None, builders=None, schedulers=None,
+                 branches=None, message_formatter=None):
+        super().__init__('all', tags, builders, schedulers, branches, subject='', add_logs=False,
+                         add_patch=False)
+        self.diffinfo_data_name = diffinfo_data_name
+        self.formatter = message_formatter
+        if self.formatter is None:
+            self.formatter = MessageFormatterFunction(lambda ctx: ctx['message'], 'plain')
+
+    @defer.inlineCallbacks
+    def generate(self, master, reporter, key, build):
+        _, _, event = key
+        if event != 'finished':
+            return None
+
+        yield utils.getDetailsForBuild(master, build,
+                                       wantProperties=self.formatter.wantProperties,
+                                       wantSteps=self.formatter.wantSteps,
+                                       wantPreviousBuild=False,
+                                       wantLogs=self.formatter.wantLogs)
+
+        if not self.is_message_needed_by_props(build):
+            return None
+        if not self.is_message_needed_by_results(build):
+            return None
+
+        report = yield self._build_report(self.formatter, master, reporter, build)
+        return report
+
+    def _get_target_changes_by_path(self, diff_info):
+        hunks_by_path = {}
+        for file_info in diff_info:
+            hunks = [(hunk['ts'], hunk['tl']) for hunk in file_info['hunks']]
+            if not hunks:
+                continue
+            hunks.sort()
+
+            hunks_by_path[file_info['target_file']] = hunks
+        return hunks_by_path
+
+    def _is_result_usable(self, result):
+        return result.get('test_code_path') is not None and result.get('line') is not None
+
+    def _is_result_in_target_diff(self, target_changes_by_path, result):
+        changes_in_path = target_changes_by_path.get(result['test_code_path'])
+        if not changes_in_path:
+            return False
+
+        result_line = result['line']
+
+        preceding_change_i = bisect.bisect(changes_in_path, (result_line + 1, -1))
```

Comment:
Not sure if bisect is very well known module of python.

I wasn't aware of it.

Also, the fact that we are comparing tuple like this looks suspicious.

I think we need a proper helper function to compare code line intervals (with unit tests)